caiman/fiola/signal_analysis_online.py

```diff
@@ -0,0 +1,460 @@
+    if do_plot:
+        plt.figure(); plt.plot(data);plt.show()
+    pks = data[signal.find_peaks(data, height=None)[0]]
+    thresh, _, _, _ = adaptive_thresh(pks, clip=100, pnorm=0.25, min_spikes=10)
+    idx = signal.find_peaks(data, height=thresh)[0]
+    
+    # Form a template
+    if template_window > 0:
+        window_length = template_window
+    else:
+        window_length = 2 
+    window = np.int64(np.arange(-window_length, window_length + 1, 1))
+    idx = idx[np.logical_and(idx > (-window[0]), idx < (len(data) - window[-1]))]
+    PTD = data[(idx[:, np.newaxis] + window)]
+    PTA = np.median(PTD, 0)
+    
+    if template_window > 0:
+        # Template matching, second time thresholding
+        t_s = np.convolve(data, np.flipud(PTA), 'same')
+        data = t_s.copy()
+    else:
+        logging.info('skip template matching')
+        data = t.copy()
+        t_s = t.copy()
+
+    median2 = np.median(data)
+    data = data - median2
+    if robust_std:
+        std = -np.percentile(data, 25) * 2 /1.35
+    else:
+        ff1 = -data * (data < 0)
+        Ns = np.sum(ff1 > 0)
+        std = np.sqrt(np.divide(np.sum(ff1**2), Ns)) 
+        
+    # Select best threshold based on estimated false positive rate
+    if adaptive_threshold:
+        pks2 = t_s[signal.find_peaks(t_s, height=None)[0]]
+        try:
+            thresh2, falsePosRate, detectionRate, low_spikes = adaptive_thresh(pks2, clip=0, pnorm=0.5, min_spikes=10)  # clip=0 means no clipping
+            thresh_factor = thresh2 / std
+            if thresh_factor < minimal_thresh:
+                logging.info(f'Adaptive threshold factor is lower than minimal theshold: {minimal_thresh}, choose thresh factor to be {minimal_thresh}')
+                thresh_factor = minimal_thresh
+            thresh2 = thresh_factor * std
+        except:
+            logging.info('Adaptive threshold fails, automatically choose thresh factor to be 3')
+            thresh_factor = 3
+            thresh2 = thresh_factor * std
+    else:
+        raise ValueError('other methods are not implemented yet')
+
+    index = signal.find_peaks(data, height=thresh2)[0]
+    logging.info(f'final threshhold equals: {thresh2/std}')
+                
+    # plot signal, threshold, template and peak distribution
+    if do_plot:
+        plt.figure()
+        plt.plot(t)
+        plt.hlines(thresh, 0, len(t_s))
+        plt.title('signal before template matching')
+        plt.show()
+        #plt.pause(3)
+
+        plt.figure()
+        plt.plot(t_s)
+        plt.hlines(thresh2, 0, len(t_s))
+        plt.title('signal after template matching')
+        plt.show()
+        #plt.pause(5)
+        
+        plt.figure();
+        plt.plot(PTA);
+        plt.title('template')
+        plt.show()
+
+        plt.figure()
+        plt.hist(pks2, 500)
+        plt.axvline(x=thresh2, c='r')
+        plt.title('distribution of spikes')
+        plt.tight_layout()
+        plt.show()  
+        plt.pause(5)
```

Comment:
Probably fine, but I wonder if there's a better answer than an arbitrary pause of 5 seconds